src/main/flight/pid.c

```diff
@@ -437,25 +437,38 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float calcHorizonLevelStrength(void)
 // processing power that it should be a non-issue.
 STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_t *pidProfile, const rollAndPitchTrims_t *angleTrim,
                                                         float currentPidSetpoint, float horizonLevelStrength)
-{
-    const float levelAngleLimit = pidProfile->levelAngleLimit;
-    // calculate error angle and limit the angle to the max inclination
-    // rcDeflection is in range [-1.0, 1.0]
-    float angle = levelAngleLimit * getLevelModeRcDeflection(axis);
+{  
+    if(axis == FD_YAW){
+        if (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(GPS_RESCUE_MODE)){
+            currentPidSetpoint = pidRuntime.angleSetpoint[FD_YAW] * cos_approx(DECIDEGREES_TO_RADIANS(attitude.values.pitch)) * cos_approx(DECIDEGREES_TO_RADIANS(attitude.values.roll)) - pidRuntime.angleSetpoint[FD_PITCH] * sin_approx(DECIDEGREES_TO_RADIANS(attitude.values.roll)) + pidRuntime.angleSetpoint[FD_ROLL] * sin_approx(DECIDEGREES_TO_RADIANS(attitude.values.pitch));
+        }
+    } else {
+        const float levelAngleLimit = pidProfile->levelAngleLimit;
+        // calculate error angle and limit the angle to the max inclination
+        // rcDeflection is in range [-1.0, 1.0]
+        float angle = levelAngleLimit * getLevelModeRcDeflection(axis);
 #ifdef USE_GPS_RESCUE
-    angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
+        angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
 #endif
-    angle = constrainf(angle, -levelAngleLimit, levelAngleLimit);
-    const float errorAngle = angle - ((attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f);
-    if (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(GPS_RESCUE_MODE)) {
-        // ANGLE mode - control is angle based
-        const float setpointCorrection = errorAngle * pidRuntime.levelGain;
-        currentPidSetpoint = pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
-    } else {
-        // HORIZON mode - mix of ANGLE and ACRO modes
-        // mix in errorAngle to currentPidSetpoint to add a little auto-level feel
-        const float setpointCorrection = errorAngle * pidRuntime.horizonGain * horizonLevelStrength;
-        currentPidSetpoint += pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
+        angle = constrainf(angle, -levelAngleLimit, levelAngleLimit);
+        const float errorAngle = angle - ((attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f);
+        if (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(GPS_RESCUE_MODE)) {
+            // ANGLE mode - control is angle based
+            const float setpointCorrection = errorAngle * pidRuntime.levelGain;
+            pidRuntime.angleSetpoint[axis] = pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
```

Comment:
You should move this to after the correction is applied, otherwise this keeps the axis from being linear. Near 90 degrees pitch roll will be very sensitive (the roll attitude), but once the correction is applied it will not be. Basically, there is a non-linear relationship to setpointCorrection and what ends up becoming the currentPidSetpoint, and thus this filter will not quite be applied correctly. Since the levelGain is a simple scalar it is ok being here.

Reply:
@Quick-Flash Where exactly do you think the attitude filtering should be applied?  We  need it on the final errorAngle correction component in both Angle and Horizon modes, and probably shouldn't be applied over the feedforward component since that would add some unnecessary delay.

Reply:
Should be applied to the currentPidSetpoint, otherwise its just not going to stay linear.

Reply:
In the angle controller with FF I think it makes sense to filter the P term and not the FF term. @Quick-Flash is right that the transformation will create some nonlinear behaviour but only testing will show if it causes a big problem.

My gut feel is that quad angle changes will be relatively slow compared to the attitude filter so nonlinearity will be small.